common/transport/amqp/src/amqp.ts

```diff
@@ -34,12 +31,90 @@ export type GenericAmqpBaseCallback<T> = (err: Error | null, result?: T) => void
  * @param   {String}    sdkVersionString        String identifying the type and version of the SDK used.
  */
 export class Amqp {
-  private _amqp: amqp10.Client;
+  //
+  // Top of the rhea object hierarchy.
+  //
+  private _rheaContainer: Container;
+  //
+  // Rhea object used to represent the connection.  The connection object is a child of the container object.
+  // It is an event emitter.  Numerous handlers listen to manage setting up a connection and act upon error events
+  // associated with the connection.
+  //
+  private _rheaConnection: Connection;
+  //
+  // Another rhea object.  It is a child of the connection.  It is an event emitter.  All rhea links will
+  // be children of the session object.
+  //
+  private _rheaSession: Session;
+  //
+  // A dictionary to handle looking up THIS transport's (NOT rhea) link objects.  The key is the actual name of the link.
+  // The value is the link object.
+  //
   private _receivers: { [key: string]: ReceiverLink; } = {};
+  //
+  // A dictionary to handle looking up THIS transport's (NOT rhea) link objects.  The key is the actual name of the link.
+  // The value is the link object.
+  //
   private _senders: { [key: string]: SenderLink; } = {};
+  //
+  // We utilize this property to hold the callback that was provided when entering a state.  This callback
+  // is invoked when an operation is completed.  Due to the event driven nature of rhea, multiple handlers
+  // will be invoked while in a state.  We need a place to keep the callback as event progression occurs.
+  //
+  private _connectionCallback: (err: Error | AmqpError, result?: any) => void;
```

Comment:
>_connectionCallback: (err: Error | AmqpError, result?: any) => void; [](start = 10, length = 68)

This is pretty much a global variable. have you explored the possbility of passing the callback from one state to the next using arguments?

Reply:
The whole issue here to me was the event driven nature of rhea.  This was the best I could come up with in dealing with the multi event nature of how rhea works.

---
In reply to: [209129265](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209129265) [](ancestors = 209129265,209128149)

Reply:
I understand that some of these might be unavoidable given how difficult it is to pass things around from one state to the next - nevertheless I think some of those booleans are there to split a state in 2 fake states - and I think the state machine could take care of that for us. 
we can leave this for later but this is something we will *need* to revisit as it is guaranteed to come back and haunt us, and to get much worse quick if we have bugs to fix later. 
Before we check this is let's at least take some time on a whiteboard to map the FSM states and how events influence transitions between them. the FSM really is a great way to "tame" the event-based nature of Rhea and I feel there's a missed opportunity here.

---
In reply to: [209328451](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209328451) [](ancestors = 209328451,209129265,209128149)